src/thor/costmatrix.cc

```diff
@@ -890,5 +930,111 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
+    const auto& target = targets.Get(target_idx);
+    // Get the indexes where the connection occurs.
+    uint32_t connedge_idx1 = source_edgestatus_[source_idx].Get(best_connection->edgeid).index();
+    uint32_t connedge_idx2 = target_edgestatus_[target_idx].Get(best_connection->opp_edgeid).index();
+
+    // set of edges recovered from shortcuts (excluding shortcut's start edges)
+    std::unordered_set<GraphId> recovered_inner_edges;
+
+    // A place to keep the path
+    std::vector<GraphId> path_edges;
+
+    // Work backwards on the forward path
+    graph_tile_ptr tile;
+    for (auto edgelabel_index = connedge_idx1; edgelabel_index != kInvalidLabel;
+         edgelabel_index = source_edgelabel_[source_idx][edgelabel_index].predecessor()) {
+      const BDEdgeLabel& edgelabel = source_edgelabel_[source_idx][edgelabel_index];
+
+      const DirectedEdge* edge = graphreader.directededge(edgelabel.edgeid(), tile);
+      if (edge == nullptr) {
+        throw tile_gone_error_t("BidirectionalAStar::FormPath failed", edgelabel.edgeid());
+      }
+
+      if (edge->is_shortcut()) {
+        auto superseded = graphreader.RecoverShortcut(edgelabel.edgeid());
+        recovered_inner_edges.insert(superseded.begin() + 1, superseded.end());
+        std::move(superseded.rbegin(), superseded.rend(), std::back_inserter(path_edges));
+      } else
+        path_edges.push_back(edgelabel.edgeid());
+    }
+
+    // Reverse the list
+    std::reverse(path_edges.begin(), path_edges.end());
+
+    // Append the reverse path from the destination - use opposing edges
+    // The first edge on the reverse path is the same as the last on the forward
+    // path, so get the predecessor.
+    for (auto edgelabel_index = target_edgelabel_[target_idx][connedge_idx2].predecessor();
+         edgelabel_index != kInvalidLabel;
+         edgelabel_index = target_edgelabel_[target_idx][edgelabel_index].predecessor()) {
+      const BDEdgeLabel& edgelabel = target_edgelabel_[target_idx][edgelabel_index];
+      const DirectedEdge* opp_edge = nullptr;
+      GraphId opp_edge_id = graphreader.GetOpposingEdgeId(edgelabel.edgeid(), opp_edge, tile);
+      if (opp_edge == nullptr) {
+        throw tile_gone_error_t("BidirectionalAStar::FormPath failed", edgelabel.edgeid());
+      }
+
+      if (opp_edge->is_shortcut()) {
+        auto superseded = graphreader.RecoverShortcut(opp_edge_id);
+        recovered_inner_edges.insert(superseded.begin() + 1, superseded.end());
+        std::move(superseded.begin(), superseded.end(), std::back_inserter(path_edges));
+      } else
+        path_edges.emplace_back(std::move(opp_edge_id));
+    }
+
+    auto edge_itr = path_edges.begin();
+    const auto edge_cb = [&edge_itr, &path_edges]() {
+      return (edge_itr == path_edges.end()) ? GraphId{} : (*edge_itr++);
+    };
+
+    Cost new_cost{0.f, 0.f};
+    const auto label_cb = [&new_cost](const EdgeLabel& label) { new_cost = label.cost(); };
```

Comment:
we only need to update the cost when recosting the path, not path necessary. once we want to tackle https://github.com/valhalla/valhalla/issues/3920, we can do it similar to bidir a*